test_lbaas.py

```diff
@@ -0,0 +1,625 @@
+    # be ready yet)
+    retry_for(seconds=30).or_fail(
+        load_balancer.verify_backend,
+        msg='Additional listener not reachable within 30s.',
+        prober=prober,
+        backend=backend,
+        port=81,
+    )
+
+
+def test_multi_listener_multi_pool(prober, create_server, image,
+                                   create_private_network,
+                                   create_load_balancer_scenario):
+    """ Two listeners connected to their own pool of member servers each.
+
+    """
+
+    # Create a load balancer setup with one backends on a private network
+    load_balancer, listener1, pool1, (backend1, ), private_network1 = \
+        create_load_balancer_scenario(num_backends=1)
+
+    # Create an additonal backend server
+    backend2 = create_server(
+        name='backend2',
+        image=image,
+        use_public_network=False,
+        use_private_network=True,
+        jump_host=prober,
+    )
+
+    # Create a second pool
+    pool2 = load_balancer.add_pool(f'lb-pool-2', 'round_robin', 'tcp')
+
+    # create an additional backend network
+    private_network2 = create_private_network(auto_create_ipv4_subnet=True)
+
+    # Create an additional backend server for the second pool
+    setup_lbaas_backend(backend2, load_balancer, pool2, private_network2)
+
+    # Add an additional listener on Port 81 for the second pool
+    load_balancer.add_listener('listener-81', pool2, 81)
+
+    # Assert backend1 is reachable on port 80 (must already be ready)
+    assert (prober.http_get(f'http://{load_balancer.vip(4)}/hostname')
+            == backend1.name)
+
+    # Assert backend2 is reachable on port 81 (try for 20s as it might not be
+    # ready yet)
+    retry_for(seconds=30).or_fail(
+        load_balancer.verify_backend,
+        msg='Additional backend and/or listener not reachable within 30s.',
+        prober=prober,
+        backend=backend2,
+        port=81,
+    )
+
+
+def test_algo_round_robin(prober, create_load_balancer_scenario):
+    """ The round_robin balancing algorithm schedules connections in turn among
+    pool members.
+
+    """
+    # Create a load balancer setup with 3 backends on a private network
+    num_backends = 3
+    load_balancer, listener, pool, backends, private_network = \
+        create_load_balancer_scenario(num_backends=num_backends)
+
+    # Issue a request to each backend to get the round robin order
+    backend_order = [load_balancer.get_url(prober, url='/hostname')
+                     for i in range(num_backends)]
+
+    # Assert all backends got a request
+    assert len(set(backend_order)) == num_backends
+
+    # Issue 10 requests to each backend and verify round robin distribution
+    for n in range(10 * num_backends):
+        # Assert the correct backend received the request
+        backend_hit = load_balancer.get_url(prober, url='/hostname')
+        assert backend_hit == backend_order[n % num_backends]
```

Comment:
It's not possible to know which backend will be hit first. This depends on the order in which they are added to the Octavia loadbalancer. Because of how Captain does this asynchronously this is not necessary the same order the requests are sent to Captain. Been there done that...

So if we really want to assert that round robin means round robin (ie. always the same order, not only evenly distributed) I don't see a way around first discovering the order of the backends.

I don't fully understand your second approach. As far as I can see it does not ensure round robin distribution. It only ensures even backend distribution. Do you think we should only check for that?

I'm open to adding more comments to the code or to replace some parts with equivalent easier to understand code. But I currently don't see much room for improvement if we really want to ensure round robin distribution.

Reply:
I still would want to ensure round-robin distribution, so my second suggestion is not a replacement. I think it would illustrate that with round-robin, we also get even distribution. But up to you.

Reply:
I'm still not sure I really understand your suggestion. The current test does not keep record of the backend hit over the whole test. Do you suggest to add a variable to keep this record and assert n hits per backend at the end?

I was thinking about how to simplify this test the other day. One thing which came to my mind was fixing the number of backends to 3 and to replace the loop at the end with the following:
```
assert load_balancer.get_url(prober, url='/hostname') == backend_order[0]
assert load_balancer.get_url(prober, url='/hostname') == backend_order[1]
assert load_balancer.get_url(prober, url='/hostname') == backend_order[2]
```

This would only assert one round of "round_robin" instead of 10, but might be easier to understand. What do you think?

Reply:
I think that is a good idea, and I would prefer that.